### ISSUE TITLE
IE 11 Security Error

### ISSUE THREAD

**Issue by CSmandato** [aws/aws-iot-device-sdk-js#111]
When connecting to a web socket the javascript sdk specifically requires the access id and secret key to be set to empty string.  This causes the sdk to set an invalid url for the web socket which leads IE 11 to throw a security error and stops the connection completely.  Is there any way to configure the sdk or url so that it does not set an invalid web socket url?  Or possibly some other work around that does not require the user to change their IE 11 security settings?  Thanks!

**Comment by liuszeng** [comment]
Hi @CSmandato,

Thank you very much for using AWS IoT Node.js SDK.

Can you elaborate more on the issue you are seeing here? Did you see any SDK error code/logs? Are you running your own application or SDK samples in IE 11?

Thanks,
Liusu

**Comment by CSmandato** [comment]
First,  let me thank you guys for providing this stuff. It sure makes development easier.  That being said here is some additional info that should show the connection failing.

Here is an image or the security error reported in the IE11 console:

<img width="1194" alt="screen shot 2017-02-15 at 11 14 27 am" src="https://cloud.githubusercontent.com/assets/25492399/22983544/6a4eccc4-f370-11e6-843e-6b91f298bda5.png">

This is the boilerplate aws iot code that we modified to create the connection:
```
var init = function (region, identityPool, endpoint, clientIdString) {

    	if (window.isUndefined(region)) throw new Error('AWS Message Service: init : region must be defined for mqtt connection.');
    	if (window.isUndefined(identityPool)) throw new Error('AWS Message Service: init : identityPool must be defined for mqtt connection.');
    	if (window.isUndefined(endpoint)) throw new Error('AWS Message Service: init : endpoint must be defined for mqtt connection.');
    	if (window.isUndefined(clientIdString)) throw new Error ('AWS Message Service: init : client Id must be set to agent guid for connection.');
    	
    	clientId = clientIdString;
    	
    	$(document).trigger('mqservice:connecting', {
    		region: region,
    		identityPool: identityPool
    	});
    	
    	setAWSRegion(region);
    	setAWSCredentials(identityPool);
    	
        //
        // Create the AWS IoT device object.  Note that the credentials must be
        // initialized with empty strings; when we successfully authenticate to
        // the Cognito Identity Pool, the credentials will be dynamically updated.
        //
        mqttClient = AWSIoTData.device({
            
            host: endpoint,
            //
            // Set the AWS region we will operate in.
            //
            region: AWS.config.region,
            //
            // Use the clientId created earlier.
            //
            clientId: clientId,
            //
            // Connect via secure WebSocket
            //
            protocol: 'wss',
            //
            // Set the maximum reconnect time to 8 seconds; this is a browser application
            // so we don't want to leave the user waiting too long for reconnection after
            // re-connecting to the network/re-opening their laptop/etc...
            //
            maximumReconnectTimeMs: 8000,
            //
            // Enable console debugging information (optional)
            //
            debug: true,
            //
            // IMPORTANT: the AWS access key ID, secret key, and sesion token must be
            // initialized with empty strings.
            //
            accessKeyId: '',
            secretKey: '',
            sessionToken: ''
        });


        //
        // Attempt to authenticate to the Cognito Identity Pool.  Note that this
        // example only supports use of a pool which allows unauthenticated
        // identities.
        //
        var cognitoIdentity = new AWS.CognitoIdentity();
        AWS.config.credentials.get(function (err, data) {
            if (!err) {
                // console.log('retrieved identity: ' + AWS.config.credentials.identityId);
                var params = {
                    IdentityId: AWS.config.credentials.identityId
                };
                cognitoIdentity.getCredentialsForIdentity(params, function (err, data) {
                    if (!err) {

                        //
                        // Update our latest AWS credentials; the MQTT client will use these
                        // during its next reconnect attempt.
                        //
                        mqttClient.updateWebSocketCredentials(
                            data.Credentials.AccessKeyId,
                            data.Credentials.SecretKey,
                            data.Credentials.SessionToken);
                    } else {
                        console.log('error retrieving credentials: ' + err);
                        throw new Error('AWS Message Service: init: Error retrieving credentials: ' + err);
                    }
                });
            } else {
                console.log('error retrieving identity:' + err);
                throw new Error('AWS Message Service: init: Error retrieving identity: ' + err);
            }
        });

        mqttClient.on('connect', mqttClientConnectHandler);
        mqttClient.on('reconnect', mqttClientReconnectHandler);
        mqttClient.on('message', mqttClientMessageHandler);
        mqttClient.on('close', mqttClientCloseHandler);
        mqttClient.on('offline', mqttClientOfflineHandler);
        mqttClient.on('error', mqttClientErrorHandler);
    };
```

We then connect to mqtt via websocket with a call similar

// AwsMessageService contains the sdk boiler plate code
var messageService = new AwsMessageService() : 

MqttSDK.init(region, identityPool, endpoint, clientIdString);



Here is the link to an issue similar to ours.  The work around provided works for us, however, we cannot ask our client to alter their IE11 configurations.

https://github.com/qzind/qz-print/issues/39

If you want to reproduce the error just bundle the sdk with your browserize script.  Include it on a page and try to make a connection to mqtt via websocket with the init method in IE11.

**Comment by liuszeng** [comment]
Hi @CSmandato ,

Thank you for providing the information.

As part of the steps of using Cognito to get temporary IAM credentials to connect to AWS IoT using MQTT over WebSocket, the SDK requires the initial value for AWS credentials to be empty string so that reconnect will be triggered and SDK credentials can be updated to use the ones retrieved via Cognito. See code line here:
https://github.com/aws/aws-iot-device-sdk-js/blob/fca6f8c52f5870957dac0f9c683e023251bec66f/device/index.js#L510

One work around would be, instead of setting the accessKeyId, secretKey and sessionToken to empty strings, give fake values to those inputs that will always result in reconnect. In this case, the logic would be same (connect with invalid creds -> failure -> reconnect -> get creds from Cognito -> update creds -> connected), except that, on the very first connect attempt, SDK will follow the same procedure to generate a "valid" (format-wise) url that could be "acceptable" by the browser.

Another work around would be, get the credentials from Cognito from the very beginning and put the credentials as part of the inputs to initialize the SDK client. Update the credentials when there is a disconnect (we are using temporary credentials here, it will get expired) or there is page refresh. In this case, the logic would be different: 
get creds from Cognito -> connect -> connected -> creds expired/refresh page -> get creds from Cognito -> update creds -> connected

Hope the above helps.

Thanks,
Liusu

**Comment by CSmandato** [comment]
Thanks for your timely reply.  I ended up setting the accessKeyId and secretKey to invalid values (essentially just a single letter for each) to by pass the wss://no-valid-credentials issue.  But that was only half the deal.  The sdk would continually connect, hold for a few seconds, then disconnect and try to reconnect.  I was able to solve that by setting the 'baseReconnectTimeMs' on the mqttClient configuration to 1000.  I don't think it matters what the value is...so long as it's set.  Now the connection is made and holds.  Hopefully this can help others who may run into the same issue.  Once again, thanks for the help!
